exporter/prometheusexporter/collector.go

```diff
@@ -0,0 +1,537 @@
+	if m.AggregationTemporality() != pdata.AggregationTemporalityCumulative {
+		return
+	}
+
+	dps := m.DataPoints()
+	for i := 0; i < dps.Len(); i++ {
+		ip := dps.At(i)
+
+		ts := pdata.UnixNanoToTime(ip.Timestamp())
+		labelValues := collectLabelValues(ip.LabelsMap(), lk)
+		signature := metricSignature(c.config.Namespace, metric, append(lk.keys, labelValues...))
+
+		indicesMap := make(map[float64]int)
+		buckets := make([]float64, 0, len(ip.BucketCounts()))
+		for index, bucket := range ip.ExplicitBounds() {
+			if _, added := indicesMap[bucket]; !added {
+				indicesMap[bucket] = index
+				buckets = append(buckets, bucket)
+			}
+		}
+		sort.Float64s(buckets)
+
+		cumCount := uint64(0)
+
+		points := make(map[float64]uint64)
+		for _, bucket := range buckets {
+			index := indicesMap[bucket]
+			var countPerBucket uint64
+			if len(ip.ExplicitBounds()) > 0 && index < len(ip.ExplicitBounds()) {
+				countPerBucket = ip.BucketCounts()[index]
+			}
+			cumCount += countPerBucket
+			points[bucket] = cumCount
+		}
+
+		c.mu.Lock()
+		v, ok := c.registeredMetrics[signature]
+		if !ok {
+			v = &metricValue{desc: desc, value: 0, labelValues: labelValues, isHistogram: true}
+			c.registeredMetrics[signature] = v
+		}
+		c.mu.Unlock()
+
+		if v.timestamp.Sub(ts) > 0 {
+			continue
+		}
+		n++
+
+		v.histogramPoints = points
+		v.histogramSum = ip.Sum()
+		v.histogramCount = ip.Count()
+		v.timestamp = ts
+
+		v.updated = time.Now()
+	}
+	return
+}
+
+/*
+	Reporting
+*/
+func (c *collector) Collect(ch chan<- prometheus.Metric) {
+	c.logger.Debug("collect called")
+
+	metrics := make([]prometheus.Metric, 0, len(c.registeredMetrics))
+
+	c.mu.Lock()
+
+	for k, v := range c.registeredMetrics {
+		var m prometheus.Metric
+		var err error
+		if v.isHistogram {
+			m, err = prometheus.NewConstHistogram(v.desc, v.histogramCount, v.histogramSum, v.histogramPoints, v.labelValues...)
+		} else {
+			m, err = prometheus.NewConstMetric(v.desc, v.metricType, v.value, v.labelValues...)
+		}
+		if err == nil {
+			if !c.config.SendTimestamps {
+				metrics = append(metrics, m)
+			} else {
+				metrics = append(metrics, prometheus.NewMetricWithTimestamp(v.timestamp, m))
+			}
+		}
+
+		if time.Now().UnixNano()-v.updated.UnixNano() > c.config.MetricExpiration.Nanoseconds() {
+			c.logger.Debug(fmt.Sprintf("metric expired: %s", m.Desc().String()))
+			delete(c.registeredMetrics, k)
+		}
+	}
+
+	c.mu.Unlock()
+
+	for _, m := range metrics {
+		ch <- m
+		c.logger.Debug(fmt.Sprintf("metric served: %s", m.Desc().String()))
+	}
+}
+
+/*
+	Helpers
+*/
+
+func metricSignature(namespace string, metric pdata.Metric, keys []string) string {
+	var b strings.Builder
+	b.WriteString(metricName(namespace, metric))
+	for _, k := range keys {
+		b.WriteString("-" + k)
+	}
+	return b.String()
+}
+
+func metricName(namespace string, metric pdata.Metric) string {
+	if namespace != "" {
+		return sanitize(namespace) + "_" + sanitize(metric.Name())
```

Comment:
I see that namespace is always coming from config. Probably that can be sanitized in advance to avoid unnecessary work.